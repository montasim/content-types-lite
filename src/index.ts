/**
 * @fileoverview Defines the various content types used in the application.
 * This module exports an object that contains different content types
 * the application commonly handles. These content types are used to
 * specify the nature of the content being transmitted.
 *
 * @module constants/contentTypesLite.ts
 * @version 1.2.0
 * @license CC BY-NC-ND 4.0
 *
 * @contact Mohammad Montasim -Al- Mamun Shuvo
 * @created 2025-01-29
 * @contactEmail montasimmamun@gmail.com
 * @contactGithub https://github.com/montasim
 */

/**
 * An object representing different content types.
 * @enum {string}
 */
const contentTypesLite = Object.freeze({
    /** Text-Based Content Types */

    /** Plain text format.
     * Example: `.txt` files.
     * Use Case: Simple text files without any formatting.
     */
    PLAIN_TEXT: 'text/plain' as const,

    /** HyperText Markup Language.
     * Example: `.html` files.
     * Use Case: Web pages and web applications.
     */
    HTML: 'text/html' as const,

    /** Cascading Style Sheets.
     * Example: `.css` files.
     * Use Case: Styling web pages.
     */
    CSS: 'text/css' as const,

    /** Comma-separated values format.
     * Example: `.csv` files.
     * Use Case: Data exchange between applications, especially spreadsheets.
     */
    CSV: 'text/csv' as const,

    /** Extensible Markup Language.
     * Example: `.xml` files.
     * Use Case: Data representation and transfer.
     */
    XML: 'text/xml' as const,

    /** Markdown text format.
     * Example: `.md` files.
     * Use Case: Writing formatted text using a plain text editor.
     */
    MARKDOWN: 'text/markdown' as const,

    /** Rich Text Format.
     * Example: `.rtf` files.
     * Use Case: Text documents with simple formatting.
     */
    RICH_TEXT: 'text/richtext' as const,

    /** Structured Data Formats */

    /** JSON format.
     * Example: `.json` files.
     * Use Case: Data interchange between web applications and servers.
     */
    JSON: 'application/json' as const,

    /** XML format for application data.
     * Example: `.xml` API responses.
     * Use Case: Data interchange in web services.
     */
    XML_APP: 'application/xml' as const,

    /** YAML Ain't Markup Language.
     * Example: `.yaml` files.
     * Use Case: Configuration files and data serialization.
     */
    YAML: 'application/x-yaml' as const,

    /** URL-encoded form data.
     * Example: query strings in `application/x-www-form-urlencoded`.
     * Use Case: Form submissions in web applications.
     */
    FORM_URLENCODED: 'application/x-www-form-urlencoded' as const,

    /** Newline-delimited JSON, used in streaming APIs.
     * Example: `.ndjson` for batch API requests.
     * Use Case: Streaming large datasets.
     */
    NDJSON: 'application/x-ndjson' as const,

    /** MessagePack format, used for efficient binary serialization.
     * Example: `.msgpack` files.
     * Use Case: Efficient data interchange.
     */
    MSGPACK: 'application/x-msgpack' as const,

    /** Protocol Buffers, used in Google's data interchange format.
     * Example: `.proto` files.
     * Use Case: Efficient data serialization.
     */
    PROTOBUF: 'application/x-protobuf' as const,

    /** Multipart & Form Data */

    /** Multipart form data.
     * Used in file uploads.
     * Use Case: Uploading files via web forms.
     */
    MULTIPART_FORM_DATA: 'multipart/form-data' as const,

    /** Multipart mixed content.
     * Used to combine different content types in emails and APIs.
     * Use Case: Sending emails with attachments.
     */
    MULTIPART_MIXED: 'multipart/mixed' as const,

    /** Alternative representations of the same content.
     * Example: multipart emails.
     * Use Case: Sending emails with different content types.
     */
    MULTIPART_ALTERNATIVE: 'multipart/alternative' as const,

    /** Digest-multipart format.
     * Used for secure email digests.
     * Use Case: Sending secure email digests.
     */
    MULTIPART_DIGEST: 'multipart/digest' as const,

    /** Related multipart format.
     * Used in web services and APIs.
     * Use Case: Sending related content in web services.
     */
    MULTIPART_RELATED: 'multipart/related' as const,

    /** Signed multipart data.
     * Used in digital signatures.
     * Use Case: Sending digitally signed content.
     */
    MULTIPART_SIGNED: 'multipart/signed' as const,

    /** Encrypted multipart data.
     * Used for secure messaging.
     * Use Case: Sending encrypted content.
     */
    MULTIPART_ENCRYPTED: 'multipart/encrypted' as const,

    /** Binary & File Upload Content Types */

    /** Generic binary data.
     * Used for raw file uploads.
     * Use Case: Uploading binary files.
     */
    OCTET_STREAM: 'application/octet-stream' as const,

    /** Portable Document Format.
     * Example: `.pdf` files.
     * Use Case: Document exchange.
     */
    PDF: 'application/pdf' as const,

    /** ZIP compressed file.
     * Example: `.zip` files.
     * Use Case: Compressing multiple files.
     */
    ZIP: 'application/zip' as const,

    /** Gzip compressed file.
     * Example: `.gz` files.
     * Use Case: Compressing single files.
     */
    GZIP: 'application/gzip' as const,

    /** Microsoft Word document.
     * Example: `.doc` files.
     * Use Case: Word processing documents.
     */
    MSWORD: 'application/msword' as const,

    /** Microsoft Word (XML format).
     * Example: `.docx` files.
     * Use Case: Word processing documents.
     */
    DOCX: 'application/vnd.openxmlformats-officedocument.wordprocessingml.document' as const,

    /** Microsoft Excel spreadsheet.
     * Example: `.xls` files.
     * Use Case: Spreadsheet documents.
     */
    EXCEL: 'application/vnd.ms-excel' as const,

    /** Microsoft Excel (XML format).
     * Example: `.xlsx` files.
     * Use Case: Spreadsheet documents.
     */
    XLSX: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' as const,

    /** Microsoft PowerPoint presentation.
     * Example: `.ppt` files.
     * Use Case: Presentation documents.
     */
    POWERPOINT: 'application/vnd.ms-powerpoint' as const,

    /** Microsoft PowerPoint (XML format).
     * Example: `.pptx` files.
     * Use Case: Presentation documents.
     */
    PPTX: 'application/vnd.openxmlformats-officedocument.presentationml.presentation' as const,

    /** Media Content Types */

    /** GIF image format.
     * Example: `.gif` files.
     * Use Case: Animated images.
     */
    GIF: 'image/gif' as const,

    /** PNG image format.
     * Example: `.png` files.
     * Use Case: Lossless image compression.
     */
    PNG: 'image/png' as const,

    /** JPEG image format.
     * Example: `.jpg`, `.jpeg` files.
     * Use Case: Lossy image compression.
     */
    JPEG: 'image/jpeg' as const,

    /** WebP image format.
     * Example: `.webp` files.
     * Use Case: Modern image format with both lossy and lossless compression.
     */
    WEBP: 'image/webp' as const,

    /** Scalable Vector Graphics.
     * Example: `.svg` files.
     * Use Case: Vector images.
     */
    SVG: 'image/svg+xml' as const,

    /** High-Efficiency Image Format.
     * Example: `.heif` files.
     * Use Case: High-efficiency image compression.
     */
    HEIF: 'image/heif' as const,

    /** AVIF image format.
     * Example: `.avif` files.
     * Use Case: High-efficiency image compression.
     */
    AVIF: 'image/avif' as const,

    /** Streaming & Event Content Types */

    /** Server-Sent Events streaming format.
     * Example: `.sse` events.
     * Use Case: Real-time data streaming.
     */
    EVENT_STREAM: 'text/event-stream' as const,

    /** TAR archive format.
     * Example: `.tar` files.
     * Use Case: Archiving multiple files.
     */
    TAR: 'application/x-tar' as const,

    /** Bzip2 compressed file.
     * Example: `.bz2` files.
     * Use Case: Compressing single files.
     */
    BZIP2: 'application/x-bzip2' as const,
});

/**
 * Type definition for available content types.
 */
export type ContentType = keyof typeof contentTypesLite;

export default contentTypesLite;
