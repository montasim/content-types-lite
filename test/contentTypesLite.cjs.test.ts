import { ContentType } from '../src';

/**
 * @fileoverview Unit tests for the contentTypesLite module.
 * These tests ensure that all content types are correctly defined, immutable, and return expected values.
 * It also validates the TypeScript types and checks that invalid keys are not allowed.
 *
 * @module test/contentTypesLite.test
 * @version 1.1.0
 * @license CC BY-NC-ND 4.0
 *
 * @contact Mohammad Montasim -Al- Mamun Shuvo
 * @created 2025-01-29
 * @contactEmail montasimmamun@gmail.com
 * @contactGithub https://github.com/montasim
 */

// eslint-disable-next-line @typescript-eslint/no-require-imports
const contentTypesLite = require('../src/index.cjs');

describe('contentTypesLite', () => {
    /**
     * Test to ensure the contentTypesLite object is defined and frozen.
     */
    it('should be defined and frozen', () => {
        expect(contentTypesLite).toBeDefined();
        expect(Object.isFrozen(contentTypesLite)).toBe(true);
    });

    /**
     * Test to ensure correct content types for text-based formats.
     */
    it('should return correct content types for text-based formats', () => {
        expect(contentTypesLite.PLAIN_TEXT).toBe('text/plain');
        expect(contentTypesLite.HTML).toBe('text/html');
        expect(contentTypesLite.CSS).toBe('text/css');
        expect(contentTypesLite.CSV).toBe('text/csv');
        expect(contentTypesLite.XML).toBe('text/xml');
        expect(contentTypesLite.MARKDOWN).toBe('text/markdown');
        expect(contentTypesLite.RICH_TEXT).toBe('text/richtext');
        expect(contentTypesLite.EVENT_STREAM).toBe('text/event-stream');
    });

    /**
     * Test to ensure correct content types for structured data formats.
     */
    it('should return correct content types for structured data formats', () => {
        expect(contentTypesLite.JSON).toBe('application/json');
        expect(contentTypesLite.XML_APP).toBe('application/xml');
        expect(contentTypesLite.YAML).toBe('application/x-yaml');
        expect(contentTypesLite.FORM_URLENCODED).toBe(
            'application/x-www-form-urlencoded'
        );
        expect(contentTypesLite.NDJSON).toBe('application/x-ndjson');
        expect(contentTypesLite.MSGPACK).toBe('application/x-msgpack');
        expect(contentTypesLite.PROTOBUF).toBe('application/x-protobuf');
    });

    /**
     * Test to ensure correct content types for multipart and form data.
     */
    it('should return correct content types for multipart and form data', () => {
        expect(contentTypesLite.MULTIPART_FORM_DATA).toBe(
            'multipart/form-data'
        );
        expect(contentTypesLite.MULTIPART_MIXED).toBe('multipart/mixed');
        expect(contentTypesLite.MULTIPART_ALTERNATIVE).toBe(
            'multipart/alternative'
        );
        expect(contentTypesLite.MULTIPART_DIGEST).toBe('multipart/digest');
        expect(contentTypesLite.MULTIPART_RELATED).toBe('multipart/related');
        expect(contentTypesLite.MULTIPART_SIGNED).toBe('multipart/signed');
        expect(contentTypesLite.MULTIPART_ENCRYPTED).toBe(
            'multipart/encrypted'
        );
    });

    /**
     * Test to ensure correct content types for binary and file upload formats.
     */
    it('should return correct content types for binary and file upload formats', () => {
        expect(contentTypesLite.OCTET_STREAM).toBe('application/octet-stream');
        expect(contentTypesLite.PDF).toBe('application/pdf');
        expect(contentTypesLite.ZIP).toBe('application/zip');
        expect(contentTypesLite.GZIP).toBe('application/gzip');
        expect(contentTypesLite.MSWORD).toBe('application/msword');
        expect(contentTypesLite.DOCX).toBe(
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document'
        );
        expect(contentTypesLite.EXCEL).toBe('application/vnd.ms-excel');
        expect(contentTypesLite.XLSX).toBe(
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
        );
        expect(contentTypesLite.POWERPOINT).toBe(
            'application/vnd.ms-powerpoint'
        );
        expect(contentTypesLite.PPTX).toBe(
            'application/vnd.openxmlformats-officedocument.presentationml.presentation'
        );
    });

    /**
     * Test to ensure correct content types for media formats.
     */
    it('should return correct content types for media formats', () => {
        expect(contentTypesLite.GIF).toBe('image/gif');
        expect(contentTypesLite.PNG).toBe('image/png');
        expect(contentTypesLite.JPEG).toBe('image/jpeg');
        expect(contentTypesLite.WEBP).toBe('image/webp');
        expect(contentTypesLite.SVG).toBe('image/svg+xml');
        expect(contentTypesLite.HEIF).toBe('image/heif');
        expect(contentTypesLite.AVIF).toBe('image/avif');
    });

    /**
     * Test to ensure correct content types for streaming and event formats.
     */
    it('should return correct content types for streaming and event formats', () => {
        expect(contentTypesLite.TAR).toBe('application/x-tar');
        expect(contentTypesLite.BZIP2).toBe('application/x-bzip2');
    });

    /**
     * Test to ensure only valid keys are allowed as ContentType type.
     */
    it('should allow only valid keys as ContentType type', () => {
        const validContentType: ContentType = 'JSON';
        expect(contentTypesLite[validContentType]).toBe('application/json');

        // @ts-expect-error This should throw a TypeScript error during compilation
        const invalidContentType: ContentType = 'INVALID';
        // Uncomment below to test runtime behavior if needed:
        expect(contentTypesLite[invalidContentType]).toBeUndefined();
    });

    /**
     * Test to match snapshot for the entire constants object.
     */
    it('should match snapshot for the entire constants object', () => {
        expect(contentTypesLite).toMatchSnapshot();
    });
});
