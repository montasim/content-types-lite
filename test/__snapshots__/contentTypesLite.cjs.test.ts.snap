// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contentTypesLite should match snapshot for the entire constants object 1`] = `
{
  "AVIF": "image/avif",
  "BZIP2": "application/x-bzip2",
  "CSS": "text/css",
  "CSV": "text/csv",
  "DOCX": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
  "EVENT_STREAM": "text/event-stream",
  "EXCEL": "application/vnd.ms-excel",
  "FORM_URLENCODED": "application/x-www-form-urlencoded",
  "GIF": "image/gif",
  "GZIP": "application/gzip",
  "HEIF": "image/heif",
  "HTML": "text/html",
  "JPEG": "image/jpeg",
  "JSON": "application/json",
  "MARKDOWN": "text/markdown",
  "MSGPACK": "application/x-msgpack",
  "MSWORD": "application/msword",
  "MULTIPART_ALTERNATIVE": "multipart/alternative",
  "MULTIPART_DIGEST": "multipart/digest",
  "MULTIPART_ENCRYPTED": "multipart/encrypted",
  "MULTIPART_FORM_DATA": "multipart/form-data",
  "MULTIPART_MIXED": "multipart/mixed",
  "MULTIPART_RELATED": "multipart/related",
  "MULTIPART_SIGNED": "multipart/signed",
  "NDJSON": "application/x-ndjson",
  "OCTET_STREAM": "application/octet-stream",
  "PDF": "application/pdf",
  "PLAIN_TEXT": "text/plain",
  "PNG": "image/png",
  "POWERPOINT": "application/vnd.ms-powerpoint",
  "PPTX": "application/vnd.openxmlformats-officedocument.presentationml.presentation",
  "PROTOBUF": "application/x-protobuf",
  "RICH_TEXT": "text/richtext",
  "SVG": "image/svg+xml",
  "TAR": "application/x-tar",
  "WEBP": "image/webp",
  "XLSX": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  "XML": "text/xml",
  "XML_APP": "application/xml",
  "YAML": "application/x-yaml",
  "ZIP": "application/zip",
}
`;
